esp32/machine_pwm: handle multiple timers.

p1ng0o opened the pull request:
This commit allow to use all the available timer for the ledc subsystem,
without affecting the API.

If a new frequency is set, first it check if an other timer is using
the same freq.
If yes, then it use this timer, otherwise it create a new one.

If all timers are used, the user should set an already used frequency,
 or de-init a channel.

This PR is related to the work done in #3608, but without the API changes.

Tested on esp32-wrover.

Commit:
esp32/machine_pwm: handle multiple timers.

This commit allow to use all the available timer for the ledc subsystem,
without affecting the API.

If a new frequency is set, first it check if an other timer is using
the same freq.
If yes, then it use this timer, otherwise it create a new one.

If all timers are used, the user should set an already used frequency,
 or de-init a channel.

dpgeorge commented:
Thanks for the contribution.

I understand that PWM on esp32 is lacking because it is fixed to LEDC_TIMER_1 and cannot have independent frequencies on different pins, so an improvement like this is welcome.  Compared to #3608 it does not change the API, as described above.

Ideally I want to work towards a `machine.PWM` API which is consistent across all ports, and this PR is compatible with that goal (at least more compatible than #3608) because it doesn't change the existing API.  So I'd be happy to merge it once a few things are fixed with it.

dpgeorge commented on ports/esp32/machine_pwm.c:
```diff
@@ -95,16 +106,60 @@ STATIC int set_freq(int newval) {
+    // Find a free PWM Timer using the same freq
+    for (timer = 0; timer < LEDC_TIMER_MAX; ++timer) {
+        if (timers[timer].freq_hz == freq) {
+            break;
```
could just `return timer` here

dpgeorge commented on ports/esp32/machine_pwm.c:
```diff
@@ -95,16 +106,60 @@ STATIC int set_freq(int newval) {
+            break;
+        }
+        if ((timer_found == -1) && (timers[timer].freq_hz == -1)) {
+            timer_found = timer;
```
could just `return timer` here

dpgeorge commented on ports/esp32/machine_pwm.c:
```diff
@@ -95,16 +106,60 @@ STATIC int set_freq(int newval) {
+            timer_found = timer;
+        }
+    }
+    if (timer >= LEDC_TIMER_MAX) {
```
could just `return -1` here if above suggestions are used

dpgeorge commented on ports/esp32/machine_pwm.c:
```diff
@@ -151,38 +206,47 @@ STATIC void esp32_pwm_init_helper(esp32_pwm_obj_t *self,
+    int freq = args[ARG_freq].u_int;
+
+    if (freq == -1) {
+        freq = PWFREQ;
```
If `freq` is not passed in as an argument (so it's -1) then the frequency should not be reconfigured (that's the existing behaviour of this function).  Eg:
```
pwm = machine.PWM(machine.Pin(0), freq=500)
pwm.init(duty=200) # this should retain the existing freq setting
```

dpgeorge replied:
I think this can be fixed by:
```
if (chan_timer[self->channel] == -1) {
    freq = PWFREQ;
} else {
    freq =  timers[chan_timer[self->channel]].freq_hz;
}
```

p1ng0o commented:
Hi @dpgeorge, thanks for the review.
Regarding the found timer algorith, I will submit an update. But we have to use an existing channel freq prior to available channel.

Regarding the init, I think I kept the behavior you described.

It freq is not passed in as an argument, I just set the freq with the default value. And that was the case before. (channel struct init).

The main difference, is that not reuse the "global" frequencie.

Not sure about this one. Let me know.

dpgeorge commented:
Before this patch the behaviour was:
```
MicroPython v1.12-683-g5fb276de3-dirty on 2020-08-26; TinyPICO with ESP32-PICO-D4
Type "help()" for more information.
>>> import machine
>>> pwm = machine.PWM(machine.Pin(0), freq=500)
>>> pwm
PWM(0, freq=500, duty=512)
>>> pwm.init(duty=200)
>>> pwm
PWM(0, freq=500, duty=200)
```
With the patch running the same code gives:
```
MicroPython v1.12-689-g7351f537e on 2020-08-27; TinyPICO with ESP32-PICO-D4
Type "help()" for more information.
>>> import machine
>>> pwm = machine.PWM(machine.Pin(0), freq=500)
>>> pwm
PWM(0, freq=500, duty=512)
>>> pwm.init(duty=200)
>>> pwm
PWM(0, freq=5000, duty=200)
```
Note the last line: the freq changed back to 5000 for this patch.

p1ng0o commented:
Understood. I will work on it. Thanks.

dpgeorge commented:
Thanks for updating, it looks good now, is this ready to merge?  (the CI failures can be ignored)

dpgeorge commented:
Actually, doing some tests, it seems that, even with only 1 active PWM, changing the frequency while it's active does not work.  Eg:
```
pwm = machine.PWM(machine.Pin(4))
pwm.freq(4000)
```
The pin stops outputing a PWM signal.

dpgeorge commented on ports/esp32/machine_pwm.c:
```diff
@@ -239,14 +303,30 @@ STATIC mp_obj_t esp32_pwm_deinit(mp_obj_t self_in) {
     // set
     int tval = mp_obj_get_int(args[1]);
-    if (!set_freq(tval)) {
+    cleanup_timer(self->channel, chan_timer[self->channel]);
```
I think it's important to make it so that there is a smooth transition to another frequency, if the same timer that's currently connected to the pin/channel can be reused.  Eg, when freq changes, if the timer on this pin is not used by any other PWM channel, and there is no other timer with the requested freq, then just call `set_freq()` on the current timer.

Commit:
esp32/machine_pwm: re-use existing timer if possible.

In order to smooth the pwm transition. If the timer is only use by one channel just set the frequency.

p1ng0o commented:
I think all you valuable review has been taken in account. I made some tests with the oscilloscope today. It seems pretty ok. But It would be better if we have some tests suite around that. Whats will be the best way to do that ? (it could be an other PR)

dpgeorge commented:
With the latest changes it seems that there is an error when trying to increase the frequency on a pin.  Eg:
```
>>> import machine
>>> p4 = machine.PWM(machine.Pin(4))
>>> p4.freq(10000)
E (326020) ledc: ledc_bind_channel_timer(174): timer_select argument is invalid
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: Failed to bind timer to channel
```

IhorNehrutsa commented:
Are any problems still exist with this PR?
I have tested it with 4 different frequencies at 8 different channels with different duties.
Frequencies and duties are right on proper pins.

p1ng0o commented:
@dpgeorge, I think this PR is mature now. Can I do something to move forward on it ?

IhorNehrutsa commented:
@p1ng0o I made a little fix to your work in #6654, but unfortunately, I lost your authority in commits, sorry.
It will be right to approve this PR and then #6654(with #6638).

IhorNehrutsa commented:
@p1ng0o, Could you test the #6654 and #6639? And what you think about #6638? Thanks. Sorry for the offtopic.

IhorNehrutsa commented on ports/esp32/machine_pwm.c:
```diff
@@ -227,6 +281,14 @@ STATIC mp_obj_t esp32_pwm_deinit(mp_obj_t self_in) {
+            timers[chan_timer[chan]].freq_hz = -1;
+            chan_timer[chan] = -1;
+        }
+
```
chan_timer[chan] = -1; // channel now don't use a timer!!!

stinos commented on ports/esp32/machine_pwm.c:
```diff
@@ -95,16 +106,46 @@ STATIC int set_freq(int newval) {
+        }
+        if (!same_freq_only && (free_timer_found == -1) && (timers[timer].freq_hz == -1)) {
+            free_timer_found = timer;
+            // Continue to check if a channel with the same freq is in used.
```
'in used' -> 'in use' or 'used'

stinos commented on ports/esp32/machine_pwm.c:
```diff
@@ -95,16 +106,46 @@ STATIC int set_freq(int newval) {
+    return free_timer_found;
+}
+
+// return true if the timer is in use in addition to curr_channel
```
Use a capital 'R' to match the other comments?
Could maybe also use some clarification, I'm not 100% sure what 'in use in addition to curr_channel' means

stinos commented on ports/esp32/machine_pwm.c:
```diff
@@ -95,16 +106,46 @@ STATIC int set_freq(int newval) {
+            return true;
+        }
+    }
+
```
Newline not really needed

p1ng0o replied:
Not sure about that one. Consistency is not matched in this file.

stinos replied:
That's fair, but I just don't see why it's needed :)

stinos commented on ports/esp32/machine_pwm.c:
```diff
@@ -151,38 +192,49 @@ STATIC void esp32_pwm_init_helper(esp32_pwm_obj_t *self,
+
+    // Check if freq wasn't passed as an argument
+    if (freq == -1) {
+        // Check if already set, otherwise use the defaut freq
```
'defaut' -> 'default'

stinos commented on ports/esp32/machine_pwm.c:
```diff
@@ -151,38 +192,49 @@ STATIC void esp32_pwm_init_helper(esp32_pwm_obj_t *self,
+    }
+
+    timer = found_timer(freq, false);
+
```
Newline not really needed?

stinos commented on ports/esp32/machine_pwm.c:
```diff
@@ -239,16 +299,60 @@ STATIC mp_obj_t esp32_pwm_deinit(mp_obj_t self_in) {
+    int new_timer = -1;
+    bool current_in_use = is_timer_in_use(self->channel, current_timer);
+
+    // Check if an already running with the same freq is running
```
'an already running' -> 'an already running timer' ?

stinos commented on ports/esp32/machine_pwm.c:
```diff
@@ -239,16 +299,60 @@ STATIC mp_obj_t esp32_pwm_deinit(mp_obj_t self_in) {
+
+    // If no existing timer was found, and the current one is in use, then find a new one
+    if (new_timer == -1 && current_in_use) {
+        // Have to found a new timer
```
'found' -> 'find'

stinos commented on ports/esp32/machine_pwm.c:
```diff
@@ -239,16 +299,60 @@ STATIC mp_obj_t esp32_pwm_deinit(mp_obj_t self_in) {
+        if (!current_in_use) {
+            // Free the old timer
+            ledc_timer_rst(PWMODE, current_timer);
+
```
Newline needed?

stinos commented:
The commit messages should start with a capital after the colon, see other commits in the repository. So for the first commit that would be for instance:
```
esp32/machine_pwm: Re-use existing timer if possible.

In order to smooth the pwm transition.  If the timer is only used by
one channel just set the frequency.
```

stinos commented on ports/esp32/machine_pwm.c:
```diff
@@ -95,16 +106,46 @@ STATIC int set_freq(int newval) {
+    // Find a free PWM Timer using the same freq
+    for (timer = 0; timer < LEDC_TIMER_MAX; ++timer) {
+        if (timers[timer].freq_hz == freq) {
+            // A timer already use the same freq. Use it now.
```
'already use' -> 'already uses' (sorry missed this one earlier)

p1ng0o commented:
@dpgeorge how can i move forward on this work ?

IhorNehrutsa commented:
I think this PR is completely ready for approval and I see no reason to further postpone its implementation.
All that's left is to click the merge button. 
:(
But we are in a https://en.wikipedia.org/wiki/Slow_television

tve commented on ports/esp32/machine_pwm.c:
```diff
@@ -261,14 +362,14 @@ STATIC mp_obj_t esp32_pwm_duty(size_t n_args, const mp_obj_t *args) {
-    duty >>= PWRES - timer_cfg.duty_resolution;
+    duty >>= PWRES - timers[chan_timer[self->channel]].duty_resolution;
     ledc_set_duty(PWMODE, self->channel, duty);
     ledc_update_duty(PWMODE, self->channel);
```
Why do you no longer check the error code on these two calls?

tve reviewed (commented):
I see two issues: one is that the error code on esp-idf calls are not checked (all it takes is wrapping the call in check_esp_err) and that some initialization routine needs to be tied into the soft reset in main.c so PWM isn't in a limbo state after a soft reset.

IhorNehrutsa commented:
This pull request is obsolete since
https://github.com/micropython/micropython/commit/af64c2ddbd758ab6bac0fcca94c66d89046663be#diff-71ce230991b3c3b3b928d13bd84430a2e5e5db6148c2da5497fba7280d6731a2 
Please close it.
#7817 is based on this PR.
Thanks, @p1ng0o

dpgeorge commented:
Aah, sorry for not getting back to this!

@IhorNehrutsa thanks for working on it.  Can you please explain (briefly) how your #7817 is different to this one?  Is it just that you updated it to work with latest master and the new `extmod/machine_pwm.c` bindings?

IhorNehrutsa commented:
This PR supports only LEDC_HIGH_SPEED_MODE
```
#define PWMODE (LEDC_HIGH_SPEED_MODE)
```
So it can operate only **4 timers and 8 channels**.
This PR supports only the ESP32 generic target.


#7817 supports up to **8 timers and 16 channels** ESP32, ESP32_S2,  ESP32_C2 chips.

dpgeorge commented:
Ok, thanks for the info!  I'll close this then in favour of #7817.  Thanks @p1ng0o for the initial work here.
